### PR TITLE
ts-web/core: reflect structs

### PR DESCRIPTION
**Pull request by pro-wh**
use Go reflection to walk the type system, generating typescript definitions. this will reduce the manual work we have to do when we update oasis-core

### CODE REVIEW

**Comment by pro-wh** [comment]
sample output from today:
[output.txt](https://github.com/oasisprotocol/oasis-sdk/files/6282353/output.txt)

**Comment by pro-wh** [comment]
sample output from today:
[types-hax.txt](https://github.com/oasisprotocol/oasis-sdk/files/7140387/types-hax.txt)

---
summary of what discrepancies remain

## interior of signature.Signed is missing
RootHashComputeBody
RootHashComputeResultsHeader
RootHashMessage
RootHashProposedBatch
RootHashRegistryMessage
RootHashStakingMessage
SGXConstraints
StorageReceiptBody
...

SignedEntity ok
ExecutorCommitment not ok
SignedProposedBatch not ok
Receipt ?
SignedTransaction ?
runtime VersionInfo.TEE -> sgx Constraints

## fixes and other updates
ConsensusStatus remove <- consensus_version
remove <- EpochTimeConsensusParameters
remove <- EpochTimeGenesis
GenesisDocument remove <- epochtime
NetTCPAddr <- NodeAddress
PVSSCommit shares PVSSCommitShare[] <- PVSSCommitShare
PVSSInstance commits Map<number, PVSSCommitState> <- Map<number, PVSSCommitShare>
StorageApplyBatchRequest rearrange
StorageLogEntry rearrange
StorageLogEntry multiline
StorageMetadata version number <- longnum
UpgradePendingUpgrade last_completed_stage longnum <- number
WorkerCommonStatus executor_roles Uint8Array <- number[]
WorkerCommonStatus storage_roles Uint8Array <- number[]

## not supported in converter yet
gas_costs key -> op
claims key -> claim
ConsensusLightParameters naming
comments

**Comment by pro-wh** [comment]
added the types inside signature.Signed types

added custom map key names for gas cost ops and staking accumulator claims

added special cases for a few types

next we'll need to pull in the doc comments

**Comment by pro-wh** [comment]
doc comments added. well that was a weird journey

sample output from today:
[types-hax.txt](https://github.com/oasisprotocol/oasis-sdk/files/7158234/types-hax.txt)

the diff from our manual types.ts is small enough for us to paste it in and describe the changes in a changelog

**Comment by pro-wh** [comment]
opening for review. still need to add something in CI though

**Comment by pro-wh** [comment]
suggestions to the repo admin:

make ci-test/ts-web-core-reflect required -- this checks that the reflect tool can build and run. if types.ts doesn't match, it's not a failure. this only shows the diff.

make ci-lint/lint-go-reflect required -- so that the reflect tool doesn't go out of style or whatever

File: .github/workflows/ci-lint.yml [hunk 1]
```diff
@@ -92,6 +92,20 @@ jobs:
...
           working-directory: tests/e2e
 
+  lint-go-reflect:
+    name: lint-go-reflect
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - name: Lint Go client-sdk
```
Comment (kostko) [inline .github/workflows/ci-lint.yml:102]:
```suggestion
      - name: Lint reflect-go
```

**Review by kostko — APPROVED** [review]
Oh, that's very nice for keeping these in sync, thanks! Will update the required checks once this is merged.

**Pull request merged**